File: clients/ts/signalr/src/JsonHubProtocol.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

import { HubMessage, IHubProtocol, MessageType } from "./IHubProtocol";
import { TextMessageFormat } from "./TextMessageFormat";
import { TransferFormat } from "./Transports";
import { InvocationMessage, StreamItemMessage, CompletionMessage, PingMessage, CloseMessage } from "./browser-index";

export const JSON_HUB_PROTOCOL_NAME: string = "json";

export class JsonHubProtocol implements IHubProtocol {

    public readonly name: string = JSON_HUB_PROTOCOL_NAME;
    public readonly version: number = 1;

    public readonly transferFormat: TransferFormat = TransferFormat.Text;

    public parseMessages(input: string): HubMessage[] {
        if (!input) {
            return [];
        }

        // Parse the messages
        const messages = TextMessageFormat.parse(input);
        console.log(messages);

        const hubMessages = [];
        for (const message of messages) {
            const parsedMessage = JSON.parse(message) as HubMessage;
            if (parsedMessage.type === undefined) {
                throw new Error("Invalid payload.");
            }
            switch (parsedMessage.type) {
                case MessageType.Invocation:
                    this.isInvocationMessage(parsedMessage);
                    break;
                case MessageType.StreamItem:
                    this.isStreamItemMessage(parsedMessage);
                    break;
                case MessageType.Completion:
                    this.isCompletionMessage(parsedMessage);
                    break;
                case MessageType.Ping:
                    // Single value, no need to validate
                    break;
                case MessageType.Close:
                    // All optional values, no need to validate
                    break;
                default:
                    throw new Error("Invalid message type.");
            }
            hubMessages.push(parsedMessage);
        }

        return hubMessages;
    }

    public writeMessage(message: HubMessage): string {
        return TextMessageFormat.write(JSON.stringify(message));
    }

    private isInvocationMessage(message: InvocationMessage): void {
        if (message.target === undefined) {
            throw new Error("Invalid payload for Invocation message.");
        }
    }

    private isStreamItemMessage(message: StreamItemMessage): void {
        if (message.invocationId === undefined) {
            throw new Error("Invalid payload for StreamItem message.");
        }
        if (message.item === undefined) {
            throw new Error("Invalid payload for StreamItem message.");
        }
    }

    private isCompletionMessage(message: CompletionMessage): void {
        if (message.result && message.error) {
            throw new Error("Invalid payload for Completion message.");
        }

        if (message.invocationId === undefined) {
            throw new Error("Invalid payload for Completion message.");
        }
    }
}

File: clients/java/signalr/src/main/java/JsonHubProtocol.java
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.util.ArrayList;
import java.util.List;

public class JsonHubProtocol implements HubProtocol {
    private final JsonParser jsonParser = new JsonParser();
    private final Gson gson = new Gson();
    private static final String RECORD_SEPARATOR = "\u001e";

    @Override
    public String getName() {
        return "json";
    }

    @Override
    public int getVersion() {
        return 1;
    }

    @Override
    public TransferFormat getTransferFormat() {
        return TransferFormat.Text;
    }

    @Override
    public HubMessage[] parseMessages(String payload) {
        String[] messages = payload.split(RECORD_SEPARATOR);
        List<HubMessage> hubMessages = new ArrayList<>();
        for (String splitMessage : messages) {
            // Empty handshake response "{}". We can ignore it
            if (splitMessage.equals("{}")) {
                continue;
            }

            JsonObject jsonMessage = jsonParser.parse(splitMessage).getAsJsonObject();
            String messageType = jsonMessage.get("type").toString();
            switch (messageType) {
                case "1":
                    //Invocation Message
                    String target = jsonMessage.get("target").getAsString();
                    JsonElement args = jsonMessage.get("arguments");
                    hubMessages.add(new InvocationMessage(target, new Object[] {args}));

                    break;
                case "2":
                    //Stream item
                    throw new UnsupportedOperationException("Support for streaming is not yet available");
                case "3":
                    //Completion
                    //Don't care yet
                    break;
                case "4":
                    //Stream invocation
                    //Don't care yet;
                    throw new UnsupportedOperationException("Support for streaming is not yet available");
                case "5":
                    //Cancel invocation
                    break;
                case "6":
                    //Ping
                    hubMessages.add(new PingMessage());
                    break;
                case "7":
                    // Close message
                    //Don't care yet;
                    break;
            }
        }
        return hubMessages.toArray(new HubMessage[hubMessages.size()]);
    }

    @Override
    public String writeMessage(InvocationMessage invocationMessage) {
        return gson.toJson(invocationMessage) + RECORD_SEPARATOR;
    }
}
